extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/WhenBuildTimeProfileProcessor.java

```diff
@@ -0,0 +1,90 @@
+
+import org.jboss.jandex.AnnotationInstance;
+import org.jboss.jandex.AnnotationTarget;
+import org.jboss.jandex.ClassInfo;
+import org.jboss.jandex.DotName;
+
+import io.quarkus.arc.WhenBuildTimeProfile;
+import io.quarkus.arc.processor.AnnotationsTransformer;
+import io.quarkus.arc.processor.DotNames;
+import io.quarkus.deployment.annotations.BuildProducer;
+import io.quarkus.deployment.annotations.BuildStep;
+import io.quarkus.runtime.configuration.ProfileManager;
+
+public class WhenBuildTimeProfileProcessor {
+
+    private static final DotName WHEN_BUILD_TIME_PROFILE = DotName.createSimple(WhenBuildTimeProfile.class.getName());
+
+    /**
+     * Uses {@link AnnotationsTransformer} to:
+     *
+     * Remove the {@code @Produces} annotation from a producer method or field when that member is also annotated with
+     * {@code @WhenBuildTimeProfile} and the current build profile doesn't match the one specified on the annotation
+     *
+     * Remove the scope annotation from a class bean that is also annotated with
+     * {@code @WhenBuildTimeProfile} and the current build profile doesn't match the one specified on the annotation
+     *
+     * This effectively means that a bean annotated with {@code @WhenBuildTimeProfile} is only enabled when
+     * the current profile matches that of the annotation
+     */
+    @BuildStep
+    void annotationTransformer(BeanArchiveIndexBuildItem beanArchiveIndex,
+            CustomScopeAnnotationsBuildItem scopes,
+            BuildProducer<AnnotationsTransformerBuildItem> annotationsTransformer) {
+        annotationsTransformer.produce(new AnnotationsTransformerBuildItem(new AnnotationsTransformer() {
+            @Override
+            public boolean appliesTo(AnnotationTarget.Kind kind) {
+                return kind == AnnotationTarget.Kind.METHOD || kind == AnnotationTarget.Kind.CLASS
+                        || kind == AnnotationTarget.Kind.FIELD;
+            }
+
+            @Override
+            public int getPriority() {
+                // Make sure this annotation transformer is invoked after all others
+                return 10;
+            }
+
+            @Override
+            public void transform(TransformationContext ctx) {
+                AnnotationInstance whenBuildTimeProfileInstance = getWhenBuildTimeProfileInstance(ctx);
+                if (whenBuildTimeProfileInstance == null) {
+                    return;
+                }
+
+                String profileOnInstance = whenBuildTimeProfileInstance.value().asString();
+                if (profileOnInstance.equals(ProfileManager.getActiveProfile())) {
+                    return;
+                }
+
+                if (ctx.isMethod() || ctx.isField()) {
+                    ctx.transform().remove(ann -> DotNames.PRODUCES.equals(ann.name())).done();
+                } else if (ctx.isClass()) {
+                    ctx.transform().remove(ann -> scopes.isScopeDeclaredOn(ann.target().asClass())).done();
```

Comment:
Hm, this should remove any annotation on a class that is annotated with a scope. Is that intentional? If so, then this could be replaced with something like:
```java
if (scopes.isScopeIn(ctx.getAnnotations())) {
  ctx.transform().removeAll().done();
}
```

Reply:
The intention was to remove only scope annotations, but now I see that I pulled the trigger a little to quickly on it :)

Reply:
Actually, you should also remove stereotypes. And additional bean defining annotations may declare a default scope too...

Reply:
So should we just go with `removeAll`? I wanted to avoid that if possible.

Reply:
I don't think just stereotypes will do.
If you don't want a class picked up as a bean, you'd need to remove any bean defining annotation from it. That can be scope, qualifier, stereotype, interceptor binding. See [spec](https://docs.jboss.org/cdi/spec/2.0/cdi-spec.html#bean_defining_annotations). Note that in Quarkus we can also have [additional bean defining annotation](https://quarkus.io/guides/cdi-reference#additional-bean-defining-annotations).

Reply:
So it seems that removing all the annotations is innevitable with this approach, right?

Reply:
Probably... but at that point it might be easier to instead just temper with Alternative and priority.
BTW how many build time profiles are there? This approach will make a bean exclusively available in one of them, what if you want the bean to be available in two or more profiles but not in third?

Reply:
For now I just pushed `removeAll`. I'll give the Alternatives approach a go later on.

Very good question about the priorities! There is always a single active build profile, so maybe we should also introduce the `WhenNotBuildTimeProfile`. How does that sound?

Reply:
@geoand sounds good. Or you could make `WhenBuildTimeProfile` a repeating annotation.

Reply:
What is the usual CDI strategy in "similar" cases? 
Best try and be consistent if possible.

Reply:
I don't see similarities that we could stick to here, might just be lack of caffeine though :)
Maybe @mkouba has some ideas?